sway/desktop/output.c

```diff
@@ -317,108 +306,159 @@ static void premultiply_alpha(float color[4], float opacity) {
+		}
+
+		if (view->border_bottom) {
+			if (con->parent->children->length == 1
+					&& con->parent->layout == L_VERT) {
+				memcpy(&color, colors->indicator, sizeof(float) * 4);
+			} else {
+				memcpy(&color, colors->child_border, sizeof(float) * 4);
+			}
+			premultiply_alpha(color, con->alpha);
+			box.x = con->x;
+			box.y = view->y + view->height;
+			box.width = con->width;
+			box.height = view->border_thickness;
+			scale_box(&box, output_scale);
+			render_rect(output->wlr_output, damage, &box, color);
 		}
-		premultiply_alpha(color, con->alpha);
-		box.x = con->x;
-		box.y = con->y + con->height - view->border_thickness;
-		box.width = con->width;
-		box.height = view->border_thickness;
-		scale_box(&box, output_scale);
-		render_rect(output->wlr_output, output_damage, &box, color);
 	}
+}
+
+/**
+ * Render a titlebar.
+ *
+ * Care must be taken not to render over the same pixel multiple times,
+ * otherwise the colors will be incorrect when using opacity.
+ *
+ * The height is: 1px border, 3px padding, font height, 3px padding, 1px border
+ * The left side for L_TABBED is: 1px border, 2px padding, title
+ * The left side for other layouts is: 3px padding, title
+ */
+static void render_titlebar(struct sway_output *output,
+		pixman_region32_t *output_damage, struct sway_container *con,
+		int x, int y, int width,
+		struct border_colors *colors, struct wlr_texture *title_texture,
+		struct wlr_texture *marks_texture) {
+	struct wlr_box box;
+	float color[4];
+	struct sway_view *view = con->type == C_VIEW ? con->sway_view : NULL;
+	float output_scale = output->wlr_output->scale;
+	enum sway_container_layout layout = con->parent->layout;
+	bool is_last_child =
+		con->parent->children->items[con->parent->children->length - 1] == con;
 
 	// Single pixel bar above title
 	memcpy(&color, colors->border, sizeof(float) * 4);
 	premultiply_alpha(color, con->alpha);
-	box.x = con->x;
-	box.y = con->y;
-	box.width = con->width;
+	box.x = x;
+	box.y = y;
+	box.width = width;
 	box.height = 1;
 	scale_box(&box, output_scale);
 	render_rect(output->wlr_output, output_damage, &box, color);
 
-	// Setting these makes marks and title easier
-	size_t inner_x = con->x + view->border_thickness * view->border_left;
-	size_t inner_width = con->width - view->border_thickness * view->border_left
-		- view->border_thickness * view->border_right;
-
 	// Single pixel bar below title
-	memcpy(&color, colors->border, sizeof(float) * 4);
-	premultiply_alpha(color, con->alpha);
-	box.x = inner_x;
-	box.y = view->y - 1;
-	box.width = inner_width;
+	size_t left_offset = 0, right_offset = 0;
+	bool connects_sides = false;
+	if (layout == L_HORIZ || layout == L_VERT ||
+			(layout == L_STACKED && is_last_child)) {
+		if (view) {
+			left_offset = view->border_left * view->border_thickness;
+			right_offset = view->border_right * view->border_thickness;
+			connects_sides = true;
+		}
+	}
+	box.x = x + left_offset;
+	box.y = y + config->font_height + 7;
```

Comment:
Is it possible to add the 3px padding and the 1px border as constants to know why there are 7, 6, 4 and 3 floating around?